lib/ts/recipe/multifactorauth/types.ts

```diff
@@ -0,0 +1,120 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { BaseRequest, BaseResponse } from "../../framework";
+import OverrideableBuilder from "supertokens-js-override";
+import { GeneralErrorResponse } from "../../types";
+import { SessionContainer } from "../session";
+import { SessionContainerInterface } from "../session/types";
+
+export type MFARequirement =
+    | {
+          id: string;
+      }
+    | string;
+
+export type MFARequirementList = (
+    | {
+          oneOf: MFARequirement[];
+      }
+    | {
+          allOf: MFARequirement[];
+      }
+    | MFARequirement
+)[];
+
+export type MFAClaimValue = {
+    c: Record<string, number>;
+    n: string[];
+};
+
+export type TypeInput = {
+    firstFactors?: string[];
+
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    firstFactors?: string[];
+
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    isAllowedToSetupFactor: (input: {
+        session: SessionContainer;
+        factorId: string;
+        requirementsForAuth: MFARequirementList;
+        factorsSetUpByTheUser: string[];
+        defaultRequiredFactorsForUser: string[];
+        defaultRequiredFactorsForTenant: string[];
+        completedFactors: Record<string, number>;
```

Comment:
do we need these as inputs? I don't mind it, but last time we discussed, it wasn't a part of the input - did you think of a use case for having these?

Reply:
We do not strictly need them, however I've found that factorsSetUpByTheUser can be useful if someone wants to implement a logic similar to our default `isAllowed...` logic. Basically we have all of these available wherever we are calling this (since we just called `getMFARequirements...`) and since there is only a single input object ignoring extra props is not an issue.